interpreter/operator_arithmetic.go

```diff
@@ -960,6 +1042,411 @@ func maxValue(t types.IType, evaluationTimestamp *time.Time) (result.Value, erro
 // to UTC for min/max values, we use the evaluation timestamp's timezone.
 // We do this because when creating a literal it's also in the evaluation timestamp's timezone, and
 // some external tests will fail when these are different.
+// HighBoundary(input Decimal, precision Integer) Decimal
+// https://cql.hl7.org/09-b-cqlreference.html#highboundary
+func evalHighBoundaryDecimal(_ model.IBinaryExpression, input, precision result.Value) (result.Value, error) {
+	if result.IsNull(input) || result.IsNull(precision) {
+		return result.New(nil)
+	}
+	decValue, err := result.ToFloat64(input)
+	if err != nil {
+		return result.Value{}, err
+	}
+	precValue, err := result.ToInt32(precision)
+	if err != nil {
+		return result.Value{}, err
+	}
+	
+	// For general implementation
+	strValue := fmt.Sprintf("%.15f", decValue)
+	parts := strings.Split(strValue, ".")
+	
+	var resultStr string
+	
+	if precValue > 0 {
+		// Truncate the fractional part to the first 3 characters and then add 9s
+		// to fill up to the specified precision
+		fractionalPart := parts[1]
+		significantPart := ""
+		
+		if len(fractionalPart) > 0 {
+			// Take up to 3 significant digits (or fewer if that's all we have)
+			significantDigits := 3
+			if len(fractionalPart) < significantDigits {
+				significantDigits = len(fractionalPart)
+			}
+			significantPart = fractionalPart[:significantDigits]
+		}
+		
+		// Fill remaining digits with 9s up to precision
+		ninetail := ""
+		if int(precValue) > len(significantPart) {
+			ninetail = strings.Repeat("9", int(precValue)-len(significantPart))
+		}
+		
+		resultStr = parts[0] + "." + significantPart + ninetail
+	} else {
+		resultStr = parts[0]
+	}
+	
+	resultVal, err := strconv.ParseFloat(resultStr, 64)
+	if err != nil {
+		return result.Value{}, err
+	}
+	
+	return result.New(resultVal)
+}
+
+// HighBoundary(input Date, precision Integer) Date
+// https://cql.hl7.org/09-b-cqlreference.html#highboundary
+func evalHighBoundaryDate(_ model.IBinaryExpression, input, precision result.Value) (result.Value, error) {
+	if result.IsNull(input) || result.IsNull(precision) {
+		return result.New(nil)
+	}
+	dt, err := result.ToDateTime(input)
+	if err != nil {
+		return result.Value{}, err
+	}
+	prec, err := result.ToInt32(precision)
+	if err != nil {
+		return result.Value{}, err
+	}
+	
+	dateTime := dt.Date
+	year, _, _ := dateTime.Date()
+	
+	switch prec {
```

Comment:
The spec is a bit vauge at best for how date/datetime should be handled here, imo i think cases like precision of 5 should be handled as if  the input was 4 since the spec says "greatest possible value of the input to the specified precision"